### PR TITLE
Basic Support for Multiplexers plus Mux Digital Input device

### PR DESCRIPTION
**Pull request by GioCC**
# Support for Multiplexer devices on Mobiflight Connector
_Matches firmware PR [135](https://github.com/MobiFlight/MobiFlight-FirmwareSource/pull/135) - firmware issue  [119](https://github.com/MobiFlight/MobiFlight-FirmwareSource/issues/119)_

Devices: MuxDriver and DigInMux

___For every detail see firmware PR [135](https://github.com/MobiFlight/MobiFlight-FirmwareSource/pull/135)___ as well as file `Mobiflight_Multiplexer_info.pdf`

### CODE REVIEW

**Comment by GioCC** [comment]
[THIS](https://github.com/MobiFlight/MobiFlight-Connector/commit/76b0509e9ecac649694fc4bf2f653cd3e26f1692) commit on official branch will require some modifications to more files in this PR. I have  to look for what changed (and why) and update accordingly.

**Comment by GioCC** [comment]
Build fails with an error on `MobiFlightUnitTests.csproj`, which makes no sense to me (and shouldn't have neen affected anyway): 
`D:\a\MobiFlight-Connector\MobiFlight-Connector\MobiFlightUnitTests\MobiFlightUnitTests.csproj(369,5): error MSB4025: The project file could not be loaded. The 'Content' start tag on line 192 position 6 does not match the end tag of 'ItemGroup'. Line 369, position 5.`

File: MobiFlight/InputConfig/InputShiftRegisterConfig.cs [hunk 1]
```diff
@@ -11,34 +11,34 @@ namespace MobiFlight.InputConfig
...
             if (onRelease != null) clone.onRelease = (InputAction)onRelease.Clone();
-            clone.pin = pin;
+            clone.ExtPin = ExtPin;
             return clone;
         }
 
         public new void ReadXml(System.Xml.XmlReader reader)
         {
-            pin = Convert.ToInt32(reader.GetAttribute(pin));
+            ExtPin = Convert.ToInt32(reader.GetAttribute(ExtPin));
```
Comment (DocMoebiuz) [inline MobiFlight/InputConfig/InputShiftRegisterConfig.cs:27]: This is a breaking change. Why did you make the change in this class in the first place?

File: MobiFlight/InputConfig/InputShiftRegisterConfig.cs [hunk 2]
```diff
@@ -11,34 +11,34 @@ namespace MobiFlight.InputConfig
...
             base.ReadXml(reader);
         }
 
         public new void WriteXml(System.Xml.XmlWriter writer)
         {
-            writer.WriteAttributeString("pin", pin.ToString());
+            writer.WriteAttributeString("ExtPin", ExtPin.ToString());
```
Comment (DocMoebiuz) [inline MobiFlight/InputConfig/InputShiftRegisterConfig.cs:33]: This is a breaking change.

File: Boards/arduino_mega.board.json [hunk 3]
```diff
@@ -48,7 +48,8 @@
     "MaxOutputs": 40,
     "MaxServos": 10,
     "MaxShifters": 4,
-    "MaxSteppers": 10
+    "MaxSteppers": 10,
+    "MaxDigInputMuxes": 2,
```
Comment (DocMoebiuz) [inline Boards/arduino_mega.board.json:52]: Please let's stick to long names.

MaxInputMultiplexer

I think we don't need "Dig" in the name. Would we need a specific multiplexer if it supported analog?

File: Boards/arduino_micro.board.json [hunk 4]
```diff
@@ -37,7 +37,8 @@
     "MaxOutputs": 18,
     "MaxServos": 3,
     "MaxShifters": 2,
-    "MaxSteppers": 3
+    "MaxSteppers": 3,
+    "MaxDigInputMuxes": 1,
```
Comment (DocMoebiuz) [inline Boards/arduino_micro.board.json:41]: MaxInputMultiplexer

File: Boards/arduino_uno.board.json [hunk 5]
```diff
@@ -33,7 +33,8 @@
     "MaxOutputs": 18,
     "MaxServos": 2,
     "MaxShifters": 2,
-    "MaxSteppers": 2
+    "MaxSteppers": 2,
+    "MaxDigInputMuxes": 1,
```
Comment (DocMoebiuz) [inline Boards/arduino_uno.board.json:37]: MaxInputMultiplexer

File: InShift2Mux.WinMerge [hunk 6]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (DocMoebiuz) [inline InShift2Mux.WinMerge:1]: This file has to be deleted

File: MobiFlight/Board.cs [hunk 7]
```diff
@@ -139,6 +139,11 @@ public class ModuleLimits
         /// </summary>
         public int MaxInputShifters = 0;
 
+        /// <summary>
+        /// Maximum number of Mux digital input blocks supported by the board.
+        /// </summary>
+        public int MaxDigInputMuxes = 0;
```
Comment (DocMoebiuz) [inline MobiFlight/Board.cs:145]: MaxInputMultiplexer

File: MobiFlight/Config/BaseDevice.cs [hunk 8]
```diff
@@ -9,9 +9,10 @@ namespace MobiFlight.Config
     public class BaseDevice : IConfigItem
     {
         protected DeviceType _type = DeviceType.NotSet;
+        protected bool _muxClient = false;
```
Comment (DocMoebiuz) [inline MobiFlight/Config/BaseDevice.cs:12]: Why do we need this in the base device class?
Let's move this to the InputMultiplexer class

File: MobiFlight/Config/BaseDevice.cs [hunk 9]
```diff
@@ -23,5 +24,9 @@ public DeviceType Type
         {
             get { return _type; }
         }
+        public bool isMuxClient
```
Comment (DocMoebiuz) [inline MobiFlight/Config/BaseDevice.cs:27]: Please remove this from the base class.
You can use Reflection to find that out if really needed.

What are we using this check for?

File: MobiFlight/Config/IConfigItem.cs [hunk 10]
```diff
@@ -10,6 +10,7 @@ public interface IConfigItem
         bool FromInternal(String value);
         DeviceType Type { get; }
         String Name { get; }
+        bool isMuxClient { get; }
```
Comment (DocMoebiuz) [inline MobiFlight/Config/IConfigItem.cs:13]: Remove this from the Interface.

File: MobiFlight/Config/MuxDriver.cs [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+using System.Xml.Serialization;
+
+namespace MobiFlight.Config
+{
+    // MuxDriver is not really used as a full-fledged BaseDevice (though it could really well be);
+    // it is mainly embedded in other devices which are mux clients.
+    // However, it inherits from BaseDevice because:
+    // 1 - it uses many of the features of the base class in the very same capacity
+    // 2 - this way it can easily be returned to the role of full Device, should it be required
+    public class MuxDriver: BaseDevice
```
Comment (DocMoebiuz) [inline MobiFlight/Config/MuxDriver.cs:12]: rename MultiplexerDriver

Why do we need this separation at all in the UI side of MobiFlight? It seems to be adding additional complexity.

File: MobiFlight/ExecutionManager.cs [hunk 12]
```diff
@@ -1525,17 +1525,25 @@ void mobiFlightCache_OnButtonPressed(object sender, InputEventArgs e)
             if (e.Type == DeviceType.InputShiftRegister)
             {
                 eventAction = MobiFlightInputShiftRegister.InputEventIdToString(e.Value);
-                // The inputKey gets the shifter pin added to it if the input came from a shift register
-                // This ensures caching works correctly when there are multiple pins on the same physical device
-                inputKey = inputKey + e.Pin;
+                // The inputKey gets the shifter external pin added to it if the input came from a shift register
+                // This ensures caching works correctly when there are multiple channels on the same physical device
+                inputKey = inputKey + e.ExtPin;
```
Comment (DocMoebiuz) [inline MobiFlight/ExecutionManager.cs:1530]: Why rename of the pin?

File: MobiFlight/InputConfig/DigInputMuxConfig.cs [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+
+// from DigInputMuxConfig
+
+namespace MobiFlight.InputConfig
+{
+    // Since digital inputs on a multiplexer are really just a bunch of buttons, 
+    // deriving from ButtonInputConfig saves copying over a ton of code 
+    // for reading/writing XML and executing actions and ensures
+    // its fundamental capabilities stay in sync with buttons.
+    public class DigInputMuxConfig : ButtonInputConfig
```
Comment (DocMoebiuz) [inline MobiFlight/InputConfig/DigInputMuxConfig.cs:15]: Rename class InputMultiplexerConfig

File: MobiFlight/InputConfigItem.cs [hunk 14]
```diff
@@ -22,6 +22,7 @@ public class InputConfigItem : IBaseConfigItem, IXmlSerializable, ICloneable, IC
         public const String TYPE_BUTTON = "Button";
         public const String TYPE_ENCODER = "Encoder";
         public const String TYPE_INPUT_SHIFT_REGISTER = "InputShiftRegister";
+        public const String TYPE_DIG_INPUT_MUX = "DigInputMux";
```
Comment (DocMoebiuz) [inline MobiFlight/InputConfigItem.cs:25]: rename TYPE_INPUT_MULTIPLEXER

File: MobiFlight/InputConfigItem.cs [hunk 15]
```diff
@@ -30,9 +31,8 @@ public class InputConfigItem : IBaseConfigItem, IXmlSerializable, ICloneable, IC
         public ButtonInputConfig button { get; set; }
         public EncoderInputConfig encoder { get; set; }
         public InputShiftRegisterConfig inputShiftRegister { get; set; }
-
+        public DigInputMuxConfig digInputMux { get; set; }
```
Comment (DocMoebiuz) [inline MobiFlight/InputConfigItem.cs:34]: Rename accordingly.

File: MobiFlight/MobiFlightDigInputMux.cs [hunk 16]
```diff
@@ -0,0 +1,77 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using CommandMessenger;
+
+namespace MobiFlight
+{
+    public class MobiFlightDigInputMux : IConnectedDevice
```
Comment (DocMoebiuz) [inline MobiFlight/MobiFlightDigInputMux.cs:9]: rename MobiFlightInputMultiplexer

File: MobiFlight/MobiFlightInputShiftRegister.cs [hunk 17]
```diff
@@ -12,6 +12,7 @@ public enum InputEvent
         {
             PRESS,
             RELEASE,
+            REPEAT,     // For future uses, like Buttons
```
Comment (DocMoebiuz) [inline MobiFlight/MobiFlightInputShiftRegister.cs:15]: why this change?

File: MobiFlight/MobiFlightInputShiftRegister.cs [hunk 18]
```diff
@@ -51,6 +52,10 @@ public static string InputEventIdToString(int enumId)
                 case (int)InputEvent.RELEASE:
                     eventAction = InputEvent.RELEASE.ToString();
                     break;
+
```
Comment (DocMoebiuz) [inline MobiFlight/MobiFlightInputShiftRegister.cs:55]: why combining this with this Pull Request?

File: MobiFlightUnitTests/MobiFlight/InputConfig/DigInputMuxConfigTests.cs [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using System.Xml;
+
+namespace MobiFlight.InputConfig.Tests
+{
+    [TestClass()]
+    public class DigInputMuxConfigTests
```
Comment (DocMoebiuz) [inline MobiFlightUnitTests/MobiFlight/InputConfig/DigInputMuxConfigTests.cs:14]: Cool that we have Unit Tests!

File: MobiFlightUnitTests/MobiFlight/InputConfig/InputShiftRegisterConfigTests.cs [hunk 20]
```diff
@@ -27,7 +27,7 @@ public void CloneTest()
         private InputShiftRegisterConfig generateTestObject()
         {
             InputShiftRegisterConfig o = new InputShiftRegisterConfig();
-            o.pin = 1;
+            o.ExtPin = 1;
```
Comment (DocMoebiuz) [inline MobiFlightUnitTests/MobiFlight/InputConfig/InputShiftRegisterConfigTests.cs:30]: Breaking change

File: UI/Panels/Input/DigInputMuxPanel.cs [hunk 21]
```diff
@@ -0,0 +1,284 @@
...
+                case MobiFlight.InputConfig.RetriggerInputAction.Label:
+                    panel = new MobiFlight.UI.Panels.Action.RetriggerInputPanel();
+                    if (isOnPress && _config != null && _config.onPress != null)
+                        (panel as MobiFlight.UI.Panels.Action.RetriggerInputPanel).syncFromConfig(_config.onPress as RetriggerInputAction);
+                    else if (!isOnPress && _config != null && _config.onRelease != null)
+                        (panel as MobiFlight.UI.Panels.Action.RetriggerInputPanel).syncFromConfig(_config.onRelease as RetriggerInputAction);
+
+                    break;
+
+                case MobiFlight.InputConfig.MSFS2020EventIdInputAction.Label:
```
Comment (DocMoebiuz) [inline UI/Panels/Input/DigInputMuxPanel.cs:113]: We shall not forget to Merge MSFS2020 Events and Custom Input since they were merged in the mean time

**Review by DocMoebiuz — CHANGES_REQUESTED** [review]

**Review by DocMoebiuz — APPROVED** [review]
Thank you very much. I will apply the review comments in a seperate PR.

File: Boards/arduino_mega.board.json (hunk 3 above)
Comment (GioCC) [reply Boards/arduino_mega.board.json:52]: The UI would be the same, but would just have to distinguish between an analog or digital input - definitely in the firmware. I guess a different object is not required, an attribute would be sufficient.

File: MobiFlight/Config/BaseDevice.cs (hunk 9 above)
Comment (GioCC) [reply MobiFlight/Config/BaseDevice.cs:27]: See the previous reply to line 12 of the same file.

File: MobiFlight/Config/MuxDriver.cs (hunk 11 above)
Comment (GioCC) [reply MobiFlight/Config/MuxDriver.cs:12]: Again, should have a look back, but I would say that all driver UI subpanels should necessarily refer to the same object with its own identity. What would be a sensible alternative? Otherwise incorporating the same information by repeating it over all InputMultiplexer objects makes little sense to me (beside being even more complex and calling for sync issues).

File: MobiFlight/Config/BaseDevice.cs (hunk 8 above)
Comment (GioCC) [reply MobiFlight/Config/BaseDevice.cs:12]: I should go back and check, but I seem to recall that this reference was somehow necessary in order to be able to access the Mux driver data from the device tree entries (the driver is not a device, therefore is not in the tree) AND/OR to identify what devices should carry the driver subpanel, also to find the first one of them, in which the subpanel is made editable.

**Comment by astrolivier76** [comment]
Hello everyone,

I'm a user come from Simvim and I made a MCP panel a few time ago.
Now, I would like to use Mobiflight whitout all rebuild from 0!!

So, I have read that now multplexer are supported.
I have create a new config in the app and I start to attribute my differents input.
But, I am quicky stopped because I can't to attribute all my switches, I explain:

My MCP panel use:

- 2 inputs multiplexers for all the switches (KD-22) and toggles
- 1 input multiplexer for all the encoders
- 1 output multiplexer for all the clock pin to the max7219 7 segment modules

All the led are connected on 2 cascaded shift register.

So, in Mobiflight:
1) I created my first input multiplexer (switch and toggle part 1), no problem -> OK
2) I created my second input multiplexer (switch and toggle part 2) -> OK
3) I would like to create a third input multiplexer for my encoders -> PB: Mobiflight can't to create more 2 multiplexers
4) I would like to create the output multiplexer -> impossible to choose in the option, not disponible in the drop down list.

So, could someone using such a configuration help me?

Can you tell me if in a future release the usable number of multiplexers will be increased?
Can you tell me if a multiplexer can be used as an output (for my 7 segments for example)

Thanks a lot for all your job,

Olivier

**Comment by GioCC** [comment]
Hi Olivier,  
first of all welcome to MobiFlight!

That's correct, multiplexers are supported, but with some limitations for now.

I'll try to briefly reply to your questions:  

- Right now, only 2 MPXs are supported (for digital inputs only), mainly to initially weed out possible issues in new code.  
There is indeed space for expansion (and their number will probably be increased soon), but that's currently the limit. (I could help you to overcome this limit even now, but there's also the next issue...)

- MPXs could technically be used for outputs (more precisely: to multiplex output devices) in MF, to a level that's practically 100% compatible with SimVim/RSC; it's just not implemented right now. 
But __I would suggest everyone interested to voice their request for such features__ (the Discord channel is a perfect place for that), in order to show that there's enough interest to posssibly trigger the decision to include them.

Please join the Discord channel for further questions and also to keep informed (and involved) in the new developments!
Regards
GiorgioCC

**Comment by astrolivier76** [comment]
Thanks for your response!!
It's a good news for me!!
I accept to give the help as possible but my knowledge is very modest. On
the other hand, I can test without worry if necessary.

I have already to try to connect me on the discord but I don't have find
where I can speak about multiplexers.
can you say me?

thanks a lot

Olivier
